pulser/simulation/simulation.py

```diff
@@ -435,21 +448,33 @@ def write_samples(
             if addr == "Global" and (
                 set(self.config.noise).issubset({"dephasing"})
             ):
-                samples_dict = self.samples["Global"][basis]
-                if not samples_dict:
-                    samples_dict = prepare_dict()
                 for slot in self._seq._schedule[channel]:
                     if isinstance(slot.type, Pulse):
-                        write_samples(slot, samples_dict, True)
-                self.samples["Global"][basis] = samples_dict
+                        # If SLM is off during slot, populate global samples
+                        if not affected_by_slm(slot):
+                            samples_dict = self.samples["Global"][basis]
+                            if not samples_dict:
+                                samples_dict = prepare_dict()
+                            write_samples(slot, samples_dict, True)
+                            self.samples["Global"][basis] = samples_dict
+                        # Otherwise, populate corresponding local
+                        else:
+                            samples_dict = self.samples["Local"][basis]
+                            for qubit in slot.targets:
+                                if qubit not in samples_dict:
+                                    samples_dict[qubit] = prepare_dict()
+                                write_samples(
+                                    slot, samples_dict[qubit], True, qubit
+                                )
+                            self.samples["Local"][basis] = samples_dict
```

Comment:
Yes I thought about it, this is what happened.
At first I implemented the mask as if you could turn it on and off many times mid simulation, and in that case any slot could be affected.
But then we decided to leave it just for the first pulse because the effect on superposition states was not trivial. The code was basically the same as before, but instead of having a list of times like [[t1_i, t1_f], [t2_i, t2_f], ...] now we just have one [t_i, t_f].
So I ended up using the code I had already written, where checking every slot made sense. I thought it could be useful to leave the code like this in case we eventually decide to implement turning the mask on and off many times.

Your way seems a good idea though, and it should be possible to expand beyond state preparation with minimal modifications